### PR TITLE
Fix Flux dist ckpt loading issue

### PR DESCRIPTION
**Pull request by suiyoubi**
Add dist loading in inference

> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/diffusion/models/flux/pipeline.py [hunk 1]
```diff
@@ -175,11 +177,23 @@ def load_from_pretrained(self, ckpt_path, do_convert_from_hf=True, save_converte
...
             save_converted_model_to (str, optional):
                 Path to save the converted checkpoint if `do_convert_from_hf` is True. Default is None.
+            load_dist_ckpt (bool, optional):
+                Whether to load the checkpoint from dist.ckpt format (NeMo2 checkpoint). Default is False.
 
         Logs:
             The function logs information about missing or unexpected keys during checkpoint loading.
         """
-        if do_convert_from_hf:
+        assert not do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'
```
Comment (orena1) [inline nemo/collections/diffusion/models/flux/pipeline.py]: `assert not do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'`
This should be 
`assert do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'`

Remove the not..

File: nemo/collections/diffusion/models/flux/pipeline.py [hunk 2]
```diff
@@ -685,12 +704,26 @@ def __init__(
...
+        controlnet_ckpt_path,
+        do_convert_from_hf=True,
+        save_converted_model_to=None,
+        load_dist_ckpt=False,
     ):
         '''
         Converts both flux base model and flux controlnet ckpt into NeMo format.
         '''
-        if do_convert_from_hf:
+        assert not do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'
```
Comment (orena1) [inline nemo/collections/diffusion/models/flux/pipeline.py:717]: also here, remove the not:
        assert do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'

Comment (orena1) [inline nemo/collections/diffusion/models/flux/pipeline.py]:
```suggestion
        assert do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'
```
Also here, remove not

**Review by orena1 — CHANGES_REQUESTED** [review]
looks cleaner, but small bug.

File: nemo/collections/diffusion/models/flux/pipeline.py (hunk 1 above)
Comment (suiyoubi) [reply nemo/collections/diffusion/models/flux/pipeline.py]: Thanks @orena1  you are right

File: nemo/collections/diffusion/models/flux/pipeline.py [hunk 3]
```diff
@@ -175,11 +177,23 @@ def load_from_pretrained(self, ckpt_path, do_convert_from_hf=True, save_converte
...
             save_converted_model_to (str, optional):
                 Path to save the converted checkpoint if `do_convert_from_hf` is True. Default is None.
+            load_dist_ckpt (bool, optional):
+                Whether to load the checkpoint from dist.ckpt format (NeMo2 checkpoint). Default is False.
 
         Logs:
             The function logs information about missing or unexpected keys during checkpoint loading.
         """
-        if do_convert_from_hf:
+        assert do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'
```
Comment (cuichenx) [inline nemo/collections/diffusion/models/flux/pipeline.py]: the assertion contradicts the explanation?

File: nemo/collections/diffusion/models/flux/pipeline.py [hunk 4]
```diff
@@ -685,12 +704,26 @@ def __init__(
...
+        controlnet_ckpt_path,
+        do_convert_from_hf=True,
+        save_converted_model_to=None,
+        load_dist_ckpt=False,
     ):
         '''
         Converts both flux base model and flux controlnet ckpt into NeMo format.
         '''
-        if do_convert_from_hf:
+        assert do_convert_from_hf and load_dist_ckpt, 'do_convert_from_hf and load_dist_ckpt cannot both be true'
```
Comment (cuichenx) [inline nemo/collections/diffusion/models/flux/pipeline.py]: same here

File: nemo/collections/diffusion/utils/flux_ckpt_converter.py [hunk 5]
```diff
@@ -219,8 +219,5 @@ def flux_transformer_converter(ckpt_path=None, transformer_config=None):
         new_state_dict[f'single_blocks.{str(i)}.mlp.linear_fc2.bias'] = (
             diffuser_state_dict[f'single_transformer_blocks.{str(i)}.proj_out.bias'].detach().clone()
         )
-        new_state_dict[f'single_blocks.{str(i)}.self_attention.linear_proj.bias'] = (
-            diffuser_state_dict[f'single_transformer_blocks.{str(i)}.proj_out.bias'].detach().clone()
-        )
```
Comment (cuichenx) [inline nemo/collections/diffusion/utils/flux_ckpt_converter.py:224]: just wondering, why is the bias removed?

Comment (suiyoubi) [reply nemo/collections/diffusion/utils/flux_ckpt_converter.py:224]: `single_blocks.{str(i)}.self_attention.linear_proj.bias` does not have bias: instead `single_transformer_blocks.{str(i)}.proj_out.bias` is mapped to the MLP FC2's bias. The whole mathematics and logic is a bit complex here. I can go through this offline with you

**Comment by pzelasko** [comment]
Closing this PR as "won't merge".
The following collections have been moved to separate repos in https://github.com/NVIDIA-NeMo organization: avlm, llm, multimodal, multimodal-autoregressive, vlm, speechlm, diffusion.
If you still wish to proceed with this contribution, please re-open it in the relevant repo.
